lib/Sema/ConstraintSystem.h

```diff
@@ -3076,16 +3081,81 @@ class ConstraintSystem {
       Expr *expr, Type conversionType, ContextualTypePurpose purpose,
       bool isOpaqueReturnType);
 
+  /// Convenience function to pass an \c ArrayRef to \c addJoinConstraint
+  Type addJoinConstraint(ConstraintLocator *locator,
+                         ArrayRef<std::pair<Type, ConstraintLocator *>> inputs,
+                         Optional<Type> supertype = None) {
+    return addJoinConstraint<decltype(inputs)::iterator>(
+        locator, inputs.begin(), inputs.end(), supertype, [](auto it) { return *it; });
+  }
+
   /// Add a "join" constraint between a set of types, producing the common
   /// supertype.
   ///
   /// Currently, a "join" is modeled by a set of conversion constraints to
-  /// a new type variable. At some point, we may want a new constraint kind
-  /// to cover the join.
+  /// a new type variable or a specified supertype. At some point, we may want
+  /// a new constraint kind to cover the join.
   ///
-  /// \returns the joined type, which is generally a new type variable.
+  /// \returns the joined type, which is generally a new type variable, unless there are
+  /// fewer than 2 input types or the \c supertype parameter is specified.
+  template<typename Iterator>
   Type addJoinConstraint(ConstraintLocator *locator,
-                         ArrayRef<std::pair<Type, ConstraintLocator *>> inputs);
+                         Iterator begin, Iterator end,
+                         Optional<Type> supertype,
+                         std::function<std::pair<Type, ConstraintLocator *>(Iterator)> getType) {
+    unsigned size = end - begin;
```

Comment:
This seems like a costly operator and it be using `std::distance`...

Reply:
In both of the existing template instantiations, `Iterator` is just a pointer so subtraction is cheap. We could use `enable_if` or a  `static_assert` to make sure `addJoinConstraint` is only called with random access iterators so that this doesn't silently do something expensive. What do you think?

Reply:
`enable_if` would be great so required capability is attached to the declaration.

Reply:
I realized we can just do something that's never expensive instead since we only care about sizes 0 and 1 🤦🏻‍♀️ I'm just going to change this to check if `begin == end` or `begin + 1 == end`